Remove confusing path decoding error handling switches

kciesielski opened the pull request:
This is a follow-up to https://github.com/softwaremill/tapir/issues/2811, https://github.com/softwaremill/tapir/issues/2610, and https://github.com/softwaremill/tapir/pull/2751

With this update, the `badRequestOnPathErrorIfPathShapeMatches` and `badRequestOnPathInvalidIfPathShapeMatches` flags will be entirely removed, as they seem to be more confusing than useful.
The PR also changes the default behavior, which will now be same as with `badRequestOnPathErrorIfPathShapeMatches = true` (the previous default was `false`).

// TODO more details about previous behavior

From now on, to override this default, one should use `.onDecodeFailureNextEndpoint` as described in `migrating.md`

Sorry for additional formatting changes to md files, they are forced by my editor autoformatting, but they don't affect how the documentation is displayed.

kciesielski commented:
@adamw the PR is ready for review. One more thing I'd like to add is a correct description in migrating.md about what was the old vs new behavior. How to describe the old default `badRequestOnPathErrorIfPathShapeMatches = false, badRequestOnPathInvalidIfPathShapeMatches = true`? (I marked it as TODO in the .md).

adamw commented on doc/migrating.md:
```diff
@@ -1,5 +1,12 @@
 
+## From 1.4 to 1.5
+
+- `badRequestOnPathErrorIfPathShapeMatches` and `badRequestOnPathInvalidIfPathShapeMatches` have been removed from `DefaultDecodeFailureHandler`. These flags were causing confusion and incosistencies caused by specifics of ZIO and Play backends. In Tapir 1.3, keeping defaults (`false` and `true` respectively for these flags) meant **TODO** From 1.4 tapir defaults to HTTP Bad Request (400) on all  ...
```
... meant that some path segment decoding failures (specifically, errors - when an exception has been thrown during decoding, but not for e.g. enumeration mismatches) were translated to a "no-match", meaning that the next endpoint was attempted. From 1.4, tapir defaults to a 400 Bad Request response to be sent instead, on all path decoding failures. This mean that:

adamw commented on server/zio-http-server/src/test/scala/sttp/tapir/server/ziohttp/ZioHttpServerTest.scala:
```diff
@@ -189,6 +189,37 @@ class ZioHttpServerTest extends TestSuite {
+            Unsafe.unsafe(implicit u => r.unsafe.runToFuture(test))
+          },
+          // https://github.com/softwaremill/tapir/issues/2764
+          Test("decode path failures result with 400") {
```
can this be moved to a general test case for all interpreters?

adamw replied:
this can be now removed?

kciesielski commented:
BTW @adamw I think we can now remove `.badRequestOnDecodeFailure` as well, as this is just the default behavior.

adamw commented:
True :) and rename the attribute's class to get rid of the flag

kciesielski commented:
@adamw Don't we still need that case class to support `onDecodeFailureNextEndpoint`?

adamw commented:
> @adamw Don't we still need that case class to support `onDecodeFailureNextEndpoint`?

Yes, but we don't need the boolean flag

kciesielski commented:
> > @adamw Don't we still need that case class to support `onDecodeFailureNextEndpoint`?
> 
> Yes, but we don't need the boolean flag

Ok I think I understand what you mean. I have just removed the boolean flag and renamed the case class to `OnDecodeFailureNextEndpointAttribute`

adamw commented on server/tests/src/main/scala/sttp/tapir/server/tests/ServerBasicTests.scala:
```diff
@@ -623,8 +652,8 @@ class ServerBasicTests[F[_], OPTIONS, ROUTE](
     }, {
       import DefaultDecodeFailureHandler.OnDecodeFailure._
```
unused import?
